src/engine/LocalVocab.h

```diff
@@ -114,11 +115,15 @@ class LocalVocab {
   // primary set of this `LocalVocab` remains unchanged.
   template <std::ranges::range R>
   void mergeWith(const R& vocabs) {
-    auto inserter = std::back_inserter(otherWordSets_);
     using std::views::filter;
+    // Note: Even though the `otherWordsSet_`is a hash set that filters out
+    // duplicates, we still manually filter out empty sets, because these
+    // typically don't compare equal to each other because of the`shared_ptr`
+    // semantics.
     for (const auto& vocab : vocabs | filter(std::not_fn(&LocalVocab::empty))) {
-      std::ranges::copy(vocab.otherWordSets_, inserter);
-      *inserter = vocab.primaryWordSet_;
+      otherWordSets_.insert(vocab.otherWordSets_.begin(),
+                            vocab.otherWordSets_.end());
+      otherWordSets_.insert(vocab.primaryWordSet_);
       size_ += vocab.size_;
```

Comment:
The change as is currently breaks the `size` calculation, as you may only add the size of otherWordSets that were no duplicates. Please also add a unit test for this.
Best you write a small lambda that does the insert + size adaptation for a single set, then your two `insert` calls become a `for_each` and a single call of the lambda.